module-5/hashcat/hashcat.py

```diff
@@ -0,0 +1,198 @@
+    """
+    with concurrent.futures.ThreadPoolExecutor() as executor:
+        futures = {executor.submit(attempt_password, password, hash_func, target_hash): password 
+                   for password in generate_passwords(charset, max_length)}
+
+        for future in concurrent.futures.as_completed(futures):
+            result = future.result()
+            if result:
+                return result
+
+    return None
+
+def dictionary_attack(
+    hash_func: Callable[[], 
+    hashlib.scrypt], 
+    target_hash: str, 
+    dictionary_file: str
+) -> Optional[str]:
+    """
+    Perform a dictionary attack to find the password that matches the target hash.
+
+    Args:
+        hash_func (Callable[[], hashlib.Hash]): The hash function to use.
+        target_hash (str): The target hash to crack.
+        dictionary_file (str): The file containing potential passwords.
+
+    Returns:
+        Optional[str]: The found password if one matches the target hash, otherwise None.
+    """
+    try:
+        with open(dictionary_file, 'r') as file:
+            for line in file:
+                password = line.strip()
+                hashed = hash_password(password, hash_func)
+                if hashed == target_hash:
+                    return password
+    except FileNotFoundError:
+        logging.error(f"Dictionary file '{dictionary_file}' not found.")
+    except Exception as e:
+        logging.error(f"An error occurred: {e}")
+    return None
+
+def main() -> None:
+    """
+    Main function to parse arguments and start the hash cracking process.
+    """
+    parser = argparse.ArgumentParser(description="Simple hash cracker.")
+    parser.add_argument("-m", "--mode", type=int, default=2, choices=[0, 1, 2, 3], help="Hash mode: 0=MD5, 1=SHA-1, 2=SHA-256, 3=SHA-512")
+    parser.add_argument("-a", "--attack", type=int, default=0, choices=[0, 1], help="Attack mode: 0=Brute-Force, 1=Dictionary")
+    parser.add_argument("--hash", type=str, help="Target hash (use with --hash-file)")
+    parser.add_argument("--hash-file", type=str, help="File containing target hash (use with --hash)")
+    parser.add_argument("--dictionary", "-d", type=str, help="Dictionary file for dictionary attack")
+    parser.add_argument("--max-length", "-ml", type=int, default=4, help="Maximum length for brute-force attack")
+    parser.add_argument("--charset", "-c", type=str, default=string.ascii_letters + string.digits, help="Charset for brute-force attack")
+
+    args = parser.parse_args()
+
+    target_hash = args.hash if args.hash else None
+    if args.hash_file:
+        try:
+            with open(args.hash_file, 'r') as file:
+                target_hash = file.read().strip()
+        except FileNotFoundError:
+            logging.error(f"Hash file '{args.hash_file}' not found.")
+            return
+
+    if not target_hash:
+        parser.error("No hash provided. Use --hash or --hash-file.")
```

Comment:
should be first check after the initialization of your parser in l. 162.

```suggestion
    if not target_hash:
        parser.error("No hash provided. Use --hash or --hash-file.")
    
    if args.hash_file:
        try:
            with open(args.hash_file, 'r') as file:
                target_hash = file.read().strip()
        except FileNotFoundError:
            logging.error(f"Hash file '{args.hash_file}' not found.")
            return
```